Fixes #15950 - CV Composite accepts latest version

parthaa opened the pull request:
This commit enables one to add the "latest" content view version to a
composite content view.
For example
If you had cv1 with versions 1,2,3 and cv2 with versions 1,2 and a new
composite content view ccv, you can now do the following operations
1) Add a cv component to ccv and say
"Always pick the latest version of the cv2" when ccv is published
2) Remove cv components of ccv individually
3) Controller calls for Content View components

This commit mainly container the model/hammer/api bindings related to these
operations.

parthaa commented:
hammer bindings provided by https://github.com/Katello/hammer-cli-katello/pull/463

daviddavis commented on app/lib/katello/api/mapper_extensions.rb:
```diff
@@ -5,7 +5,6 @@ def api_resources(*args, &block)
         options          = args.extract_options!
         options[:except] = Array(options[:except])
         options[:except].push(:new, :edit)
-
```
Intentional?

daviddavis commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,85 @@
+    end
+
+    api :GET, "/content_views/:composite_content_view_id/content_view_components/:id", N_("Show a content view component")
+    param :composite_content_view_id, :number, :desc => N_("composite content view numeric identifier"), :required => true
```
Why is this param required? Why can't i just show the component with just an id for it?

daviddavis replied:
I guess my point is that it would be nice to have a `/content_view_components/:id` route so I don't have to also pass the content view id every time.

daviddavis commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,85 @@
+    param :content_view_id, :identifier,
+          :desc => N_("content view identifier of the component who's latest version is desired")
+    param :latest, :bool, :desc => N_("true if the latest version of the components content view is desired")
+
```
Blank line.

daviddavis commented on app/models/katello/content_view_component.rb:
```diff
@@ -2,23 +2,80 @@ module Katello
+
     private
 
     def content_view_types
```
Probably should break up this method.

daviddavis replied:
Also, mind renaming this to something like `ensure_valid_attributes` or some such?

daviddavis commented on app/views/katello/api/v2/content_view_components/show.json.rabl:
```diff
@@ -0,0 +1,45 @@
+  child :archived_repos => :repositories do
+    attributes :id, :name, :label, :description
+  end
+end
```
Seems like these attributes can be DRYed up either by:

1) Combining content_view_version and latest_version (users should be able to look at `latest` to determine if the version is the latest version or just a content view version)
2) Or using a shared partial template (if you don't want to combine latest_version and content_view_version)

parthaa replied:
@daviddavis Good suggestion. Updated.

daviddavis commented on config/routes/api/v2.rb:
```diff
@@ -54,6 +54,17 @@ class ActionDispatch::Routing::Mapper
+        # resources :content_views, :param => :composite_content_view_id, :only => [] do
+        #   api_resources :content_view_components
+        #   api_resources :components, :only => [:index]
+        # end
```
Still needed?

daviddavis commented:
BTW, I didn't see any calls to `readable` or permissions changes. Do you know if this stuff works with a non-admin user?

daviddavis commented on app/views/katello/api/v2/content_view_components/create.json.rabl:
```diff
@@ -0,0 +1,3 @@
+object @resource
+
+extends "katello/api/v2/puppet_modules/base"
```
puppet_modules?

daviddavis commented on app/models/katello/content_view_component.rb:
```diff
@@ -2,24 +2,89 @@ module Katello
+
+    def latest_version
+      if latest?
+        self.content_view.versions.last
```
The `last` method here I think will sort by primary key. This probably won't work if the user has an incremental update (e.g. 2.2) that was published after a major update (e.g. 3). Probably need to sort by version major/minor instead.

parthaa replied:
excellent point. Thanks!

daviddavis commented:
When I removed the only version for a content view that I added via `content_view_id` and `lastest`, then I ran into all sorts of problems (I couldn't even visit the content view page in the app). Other than that edge case, everything else seemed to work fine.

daviddavis commented on app/models/katello/content_view.rb:
```diff
@@ -16,14 +16,11 @@ class ContentView < Katello::Model
+             :inverse_of => :composite_content_view, :foreign_key => :composite_content_view_id
+
+    has_many :latest_content_view_components, :source => :content_view_components, :class_name => "Katello::ContentViewComponent",
+             :dependent => :destroy, :inverse_of => :content_view, :foreign_key => :content_view_id
```
I would probably remove this association. YAGNI.

parthaa replied:
So I need to handle the case where you either remove the CV or the cv component (for the latest cv cases ).  I had this association to tackle that, because I couldn't figure out how else to remove those.

daviddavis replied:
I see. In that case, I might rename it to `lastest_component_content_views` since they are content views and not components.

daviddavis replied:
Also, I think the class_name should be ContentView and not ContentViewComponent.

parthaa commented:
> When I removed the only version for a content view that I added via content_view_id and lastest, then I ran into all sorts of problems (I couldn't even visit the content view page in the app). Other than that edge case, everything else seemed to work fine.

@daviddavis updated this edge case. Basically the component versions wont get published if the cv has no versions

parthaa commented:
> BTW, I didn't see any calls to readable or permissions changes. Do you know if this stuff works with a non-admin user?

@daviddavis good call. Fixed those in the latest commit

parthaa commented:
@daviddavis I have added the tests for content_view_component and content_view_components controller which I believe are fairly comprehensive. Kindly let me know if you have other things I need to address in this PR

daviddavis commented:
There's still a lot of duplication in the validation on component. I would probably clean it up with something like this:

https://gist.github.com/daviddavis/8ad8a806351434693b0c5a42983500d0

daviddavis commented:
Tests and changes look good. I think this'll be ready to go if you fix the failing tests and refactor the validation in component.

jlsherrill commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,91 @@
+      options[:latest] = ::Foreman::Cast.to_bool(component_params[:latest]) if component_params.key?(:latest)
+      respond resource: ContentViewComponent.create!(options)
+    end
+
```
might want to consider how the UI will handle with the current bulk add/remove

parthaa commented:
> There's still a lot of duplication in the validation on component. I would probably clean it up with something like this:
@daviddavis  updated

parthaa commented:
Added UI bindings accordingly

- Add Component Content Views
![01-add-component-content views](https://cloud.githubusercontent.com/assets/1069779/20209253/620a0e48-a7c1-11e6-848d-197d8ca3d620.png)

- List Component Content VIews
![02-list-content views](https://cloud.githubusercontent.com/assets/1069779/20209268/7bcdc1c6-a7c1-11e6-80b0-7b01ead6d054.png)


- Update the Latest Version Selection
![03-list-content views- latest](https://cloud.githubusercontent.com/assets/1069779/20209277/89843fca-a7c1-11e6-905e-9b0637b405b1.png)

- Published Outcome
![04-published-content-views- latest](https://cloud.githubusercontent.com/assets/1069779/20209291/9cb875ac-a7c1-11e6-9412-cd64af3abd22.png)

johnpmitsch commented:
Sorry for the comment after changes are approved, but I was confused if "Latest" meant use the current latest version and then keep using that same version, or to always use the latest version when new ones are published

Maybe we should consider being more verbose and saying "Always use latest" or even "Use latest"? I could see users getting confused.

bbuckingham commented:
@parthaa, in addition to @johnpmitsch's point, I'd recommend that we be consistent in wording with the puppet module selection, since it also supports selecting the latest version.

For puppet modules, it appears behavior is:
- when selecting a puppet module:
  'Use Latest (currently 3.0)'

- when viewing the puppet modules that were selected:
  'Latest (currently 3.0)'

For all other entries (not latest) we'd simply list the version number.

bbuckingham commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,100 @@
+      param :content_view_version_id, :identifier, :desc => N_("identifier of the version of the component content view")
+      param :content_view_id, :identifier,
+            :desc => N_("content view identifier of the component who's latest version is desired")
+      param :latest, :bool, :desc => N_("true if the latest version of the components content view is desired")
```
s/components/component's/

bbuckingham commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,100 @@
+    api :PUT, "/content_views/:composite_content_view_id/content_view_components/remove",
+                                        N_("Remove components from the content view")
+    param :composite_content_view_id, :identifier, :desc => N_("composite content view identifier"), :required => true
+    param :component_ids, Array, :desc => N_("Array of component ids to remove"), :required => true
```
Are these the Ids of the component content view versions or the component content view? It may be good to clarify for the user.

parthaa replied:
All the component ids in controller refer to the content_view_component id.  Its neither component's content_view id nor is it components content_view_version id. That being said I am wondering if I should make all the component_ids accept  content view component's cv id.

bbuckingham commented on app/controllers/katello/api/v2/content_view_components_controller.rb:
```diff
@@ -0,0 +1,100 @@
+      if component_params.key?(:latest) && component_params.key?(:content_view_version_id)
+        latest = ::Foreman::Cast.to_bool(component_params[:latest])
+        if latest && cvv_id.present?
+          fail HttpErrors::UnprocessableEntity,
```
Curious, do you need to confirm this here, since there is also a validation at the model level?

parthaa replied:
> @parthaa, in addition to @johnpmitsch's point, I'd recommend that we be consistent in wording with the puppet module selection, since it also supports selecting the latest version.
> 
I have updated the templates to match a combination of @johnpmitsch and @bbuckingham 's views in consultation with @ehelms 

Basically the drop down will say "Always Use Latest (Currently 2.0)" when one is updating or adding a content view.  However list will show "Latest(Currently 2.0)" similar to what puppet modules page shows. I also created a redmine issue to update the puppet UI and have it say 'Always Use Latest (Currently 1.4.0)' -> http://projects.theforeman.org/issues/17390

parthaa replied:
Code has been updated to reflect the above scenario.

parthaa replied:
@bbuckingham so I debated about this. 
3 cases to think of where we are updating an existing component.

1. params = {latest: true}
1. params = {latest: true, content_view_version_id => same_as_old_version_id }
1. params = {latest: true, content_view_version_id => different_version_id }

Now if I were to move this controller logic to model, I 'd have to add a before_validate that says
```
if latest was changed then clear out the content view version
```

but now if the request both latest=true and  had content_view_version_id we'd need to fail the save in that case. A model will not know that  a content_view_version_id got passed in the request . Since its in the request itself I believe its best suited in the controller.

Model level validation just makes sure that the params are set in a valid state before save.

bbuckingham commented:
@parthaa, Reviewed and tested.  Only minor comments on the review and didn't observe any issues during the test.  ACK from me with resolution of the minor comments.

Also, as we discussed in irc, it would be good for a core UI dev to also review the UI portions (e.g. @waldenraines ).  I reviewed them; however, I know there are others with more knowledge of that area.

Great job Partha!  This will be an awesome feature!

waldenraines commented:
The UI changes look fine and if I come across something when fixing the conflicts this will have with https://github.com/Katello/katello/pull/6460 I will fix it.

bbuckingham commented:
@parthaa, would it be possible to increase the width of the 'Version' column so that the user can see the full initial value?  I can see them if I widen the browser; however, that UI would require all of  screen real my estate. :)

In the attached screenshot, the user would essentially have to click on every row to see what is currently latest:
![version_selector](https://cloud.githubusercontent.com/assets/1319246/20526619/d0134c5e-b092-11e6-8fdf-b409b65533e2.png)

parthaa commented:
> @parthaa, would it be possible to increase the width of the 'Version' column so that the user can see the full initial value? I can see them if I widen the browser; however, that UI would require all of screen real my estate. :)

Sorry for the delay on this. Few things on this.
1)  So the drop size there is based of table cell width and we don't have stylings elsewhere where we forcibly set the width of the select box (dont think we can control it)
2) @waldenraines ' pull request might render this change unnecessary since the left nav  is going to vanish and the table will stretch more.
3) Moreover even on my laptop the drop down expanded to this -> 
![version_selector](http://i.imgur.com/cWuU4Yl.png)

bbuckingham commented:
ACK.  If the version information isn't fully visible after @waldenraines' changes go in for removal of tupane, we can create a separate usability redmine.
